src/test/SecurityPanel/StoreJmxCredentials.test.tsx

```diff
@@ -0,0 +1,222 @@
+  return {
+    CreateJmxCredentialModal: jest.fn((props) => {
+      return (
+        <Modal isOpen={props.visible} variant={ModalVariant.large} showClose={true} onClose={props.onClose} title="CreateJmxCredentialModal">
+          Jmx Auth Form
+        </Modal>
+      );
+    }),
+  };
+});
+
+jest.mock('@app/Shared/Services/Api.service', () => {
+  const mockTarget = { connectUrl: 'service:jmx:rmi://someUrl', alias: 'fooTarget' } as Target;
+const anotherMockTarget = { connectUrl: 'service:jmx:rmi://anotherUrl', alias: 'anotherTarget' } as Target;
+  return {
+    ApiService: jest.fn(() => {
+      return {
+        getTargetsWithStoredJmxCredentials: jest.fn()
+        .mockReturnValueOnce(of([mockTarget]))
+        .mockReturnValueOnce(of([]))
+        .mockReturnValueOnce(of([mockTarget]))
+        .mockReturnValueOnce(of([mockTarget]))
+        .mockReturnValueOnce(of([mockTarget, anotherMockTarget]))
+        .mockReturnValueOnce(of([mockTarget, anotherMockTarget])),
+        deleteTargetCredentials: jest.fn().mockReturnValue(of(true)),
+      };
+    }),
+  };
+});
+
+jest.mock('@app/Shared/Services/NotificationChannel.service', () => {
+  const mockNotification  = {message: {target: 'service:jmx:rmi://someUrl'}} as NotificationMessage;
+  const anotherMockNotification  = {message: {target: 'service:jmx:rmi://someUrl'}} as NotificationMessage;
+
+  return {
+    NotificationChannel: jest.fn(() => {
+      return {
+        messages: jest.fn()
+        .mockReturnValueOnce(of()) // '<StoreJmxCredentials />'
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of()) 
+        .mockReturnValueOnce(of()) // 'displays empty state text when the table is empty'
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of())  
+        .mockReturnValueOnce(of()) // 'opens the JMX auth modal when Add is clicked'
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of()) 
+        .mockReturnValueOnce(of()) // 'renders a table entry when credentials are stored'
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of()) 
+        .mockReturnValueOnce(of()) // 'removes a single table entry when the selected credentials are deleted'
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of(mockNotification))
+        .mockReturnValueOnce(of()) // 'removes all table entries when all credentials are deleted'
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of())
+        .mockReturnValueOnce(of(mockNotification, anotherMockNotification)),
+      };
+    }),
+  };
+});
+
+jest.mock('@app/Shared/Services/Target.service', () => {
+  return {
+    TargetService: jest.fn(() => {
+      return {
+        deleteCredentials: jest.fn(),
+      };
+    }),
+  };
+});
```

Comment:
@andrewazores I think I'm missing the correct service name for `TargetService` based on the error below. I've tried replacing `TargetService: jest.fn()` with `TargetInstance jest.fn()` and `target: jest.fn()` with the same results. Any ideas? 
```
● <StoreJmxCredentials /> › removes a single table entry when the selected credentials are deleted

    TypeError: Cannot read properties of undefined (reading 'deleteCredentials')

      127 |         handleRowCheck(false, idx);
      128 |         tasks.push(context.api.deleteTargetCredentials(t));
    > 129 |         context.target.deleteCredentials(t.connectUrl);
          |                        ^
      130 |       }
      131 |     });
      132 |     addSubscription(forkJoin(tasks).subscribe());

      at src/app/SecurityPanel/StoreJmxCredentials.tsx:129:24
```

Reply:
@hareetd ^ ?

Judging by the type error it looks like your `context.target` reference is undefined. So there is a `context` object in scope, but it's either a completely empty object or at least the `.target` property is undefined/null/uninitialized.

If I understand the mocking system here, I think what you've done seems correct, and by providing these mocks the real `Services.tsx` (https://github.com/cryostatio/cryostat-web/blob/main/src/app/Shared/Services/Services.tsx) should populate the `context` correctly via the `ServiceContext`, but that looks to me like the thing that's broken in this instance.

Reply:
Hmm, it looks Jest has issues when it comes to mocking exported `const` members of modules. This line is what seems to be issue: https://github.com/cryostatio/cryostat-web/blob/4a02275ee9f839bbb4d3820880e86728ae89dd88/src/app/Shared/Services/Target.service.tsx#L113. There's probably a workaround out there; I'll keep looking into this as well.

Reply:
Never mind, I figured out the real issue. Your current approach to mocking the `@app/Shared/Services/Target.service` module replaces the module definition with a mocked `TargetService` but any other exported members of the original module (i.e. `NO_TARGET`, `Target` and `TargetInstance`) are left undefined. In order to ensure these other exported members are preserved, you'll need a `jest.requireActual` call in your mock definition:

```
jest.mock('@app/Shared/Services/Target.service', () => {
  return {
    ...jest.requireActual('@app/Shared/Services/Target.service'),
    TargetService: jest.fn(() => {
      return {
        deleteCredentials: jest.fn(),
      };
    }),
  };
});
```
You would've needed to do this anyways in order to use the `Target` import in your mock of the `TargetsService`.

Reply:
Thanks! I think I still need to import the `Target` interface because I made a mock `Target` return object. The tests are compiling now.